qiskit/transpiler/__init__.py

```diff
@@ -130,14 +139,75 @@
 this can be overridden by passing explicit ``<stage>_method="<choice>"`` arguments to
 :func:`.generate_preset_pass_manager`.
 
-.. note::
 
-    The preset pass managers almost always include stochastic, heuristic-based passes.  If you need
-    to ensure reproducibility of a compilation, pass a known integer to the ``seed_transpiler``
-    argument to the generator functions.
+.. _transpiler-preset-determinism:
+
+Reproducibility of the preset pipelines
+---------------------------------------
+
+Quantum compilation often involves solving problems that are knownn to be non-polynomial in
+complexity, and so are intractable to globally optimize.  In these cases, stochastic and heuristic
+algorithms are often more appropriate.  This leads to problems of reproducibility, however.
+
+The preset pass managers almost always include stochastic, heuristic-based passes.  If you need
+to ensure reproducibility of a compilation, pass a known integer to the ``seed_transpiler``
+argument to the generator functions.
+
+All built-in plugins to Qiskit are required to produce their analyses and modify the
+:class:`.DAGCircuit` in deterministic ways if they randomization (if any) is seeded, so that a
+compilation can be repeated later.  There are limits on this:
+
+* All built-in passes with stochastic components must provide a way to seed the randomization, and
+  if seeded, they must respect the rules of deterministic output.
+
+* All built-in passes without stochastic components must respect the rules of deterministic
+  output for identical input. It is permissible to keep a cache for efficiency, but given the same
+  set of inputs, the pass's returns must be the same if the pass is called multiple times, unless
+  something outside the pass's control mutates one of its inputs in-place (for example, the
+  :class:`.BasisTranslator` uses the :data:`.SessionEquivalenceLibrary` by default in the preset
+  pass managers, and it is not a bug to get different results if new entries are added to the
+  equivalence library).  An "output" is anything the pass writes out for further consumption; this
+  can be the explicit ``return`` value from the pass, but also includes properties intended for
+  later consumption in the :class:`.PropertySet`.
+
+* The output of a pass should be deterministic on a given machine for a given version of Qiskit and
+  a frozen environment, no matter how many threads are available for the pass.  Many built-in Qiskit
+  passes use threaded concurrency, and they are not permitted to have different behavior based on
+  the number of threads.
+
+* The output of a pass for a fixed seed is not required to be equal if some part of the underlying
+  Python environment changes (such as a dependent package updating), or if the system mathematical
+  libraries change (such as a different implementation of OpenBLAS is available).
```

Comment:
Or did you mean version of `OpenBLAS`?

```suggestion
  libraries change (such as a different implementation of BLAS is available).
```